### ISSUE TITLE
Facebook as external identity provider fails with 400 Bad request

### ISSUE THREAD

**Issue by jm4rc05** [cloudfoundry/uaa#567]
Trying to connect Facebook as an external OAuth2/OIDC provider to UAA.

Relevant section in login.yml:

```yml
            facebook:
                type: oidc1.0
                authUrl: https://www.facebook.com/dialog/oauth
                tokenUrl: https://graph.facebook.com/oauth/access_token
                tokenKey: <REDACTED>
                issuer: https://www.facebook.com
                scopes:
                    - public_profile
                    - email
                    - user_friends
                linkText: Facebook
                showLinkText: true
                addShadowUserOnLogin: true
                relyingPartyId: <REDACTED>
                relyingPartySecret: <REDACTED>
                skipSslValidation: false
                attributeMappings:
                    user_name: email
```

Receiving this at UAA.log:

```log
[2017-03-18 11:53:21.461] cloudfoundry-identity-server - ???? [http-nio-8080-exec-3] .... DEBUG --- SecurityFilterChainPostProcessor$HttpsEnforcementFilter: Filter chain 'xOauthCallbackEndpointSecurity' processing request POST /uaa/login/callback/facebook
[2017-03-18 11:53:21.466] cloudfoundry-identity-server - ???? [http-nio-8080-exec-3] .... DEBUG --- XOAuthAuthenticationManager: Starting external authentication for:org.cloudfoundry.identity.uaa.provider.oauth.XOAuthCodeToken@5a3d1511
[2017-03-18 11:53:21.556] cloudfoundry-identity-server - ???? [http-nio-8080-exec-3] .... DEBUG --- XOAuthAuthenticationManager: Performing token exchange with url:https://graph.facebook.com/oauth/access_token and request:{grant_type=[authorization_code], response_type=[token], code=[PROTECTED], redirect_uri=[http://localhost:8080/uaa/login/callback/facebook]
[2017-03-18 11:53:22.191] cloudfoundry-identity-server - ???? [http-nio-8080-exec-3] .... ERROR --- XOAuthAuthenticationFilter: XOauth Authentication exception
org.springframework.web.client.HttpClientErrorException: 400 Bad Request
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:91)
	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:700)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:653)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:628)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:579)
	at org.cloudfoundry.identity.uaa.provider.oauth.XOAuthAuthenticationManager.getTokenFromCode(XOAuthAuthenticationManager.java:419)
	at org.cloudfoundry.identity.uaa.provider.oauth.XOAuthAuthenticationManager.getClaimsFromToken(XOAuthAuthenticationManager.java:334)
	at org.cloudfoundry.identity.uaa.provider.oauth.XOAuthAuthenticationManager.getExternalAuthenticationDetails(XOAuthAuthenticationManager.java:113)
	at org.cloudfoundry.identity.uaa.provider.oauth.XOAuthAuthenticationManager.getExternalAuthenticationDetails(XOAuthAuthenticationManager.java:91)
	at org.cloudfoundry.identity.uaa.authentication.manager.ExternalLoginAuthenticationManager.authenticate(ExternalLoginAuthenticationManager.java:116)
	at org.cloudfoundry.identity.uaa.provider.oauth.XOAuthAuthenticationFilter.authenticationWasSuccessful(XOAuthAuthenticationFilter.java:83)
	at org.cloudfoundry.identity.uaa.provider.oauth.XOAuthAuthenticationFilter.doFilter(XOAuthAuthenticationFilter.java:59)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
```

**Comment by jm4rc05** [comment]
I'm trying some different parameters combinations and figuring out what I did wrong and now stoped here [LoginInfoEndpoint. captureImplicitValuesUsingJavascript()](https://github.com/cloudfoundry/uaa/blob/master/server/src/main/java/org/cloudfoundry/identity/uaa/login/LoginInfoEndpoint.java#L651).

What this method is supposed to do? I'm confused!

**Comment by sreetummidi** [comment]
Hello, 
We have not tested our integration with social providers other than Google. This will definitely need some code changes to support. I already have adding support for Github and Facebook on the horizon. I will keep this issue open and we will add support for Facebook auth.

Thanks,
Sree

**Comment by jm4rc05** [comment]
Hi Sree,

Glad to read this. Do you have a time frame for us? I'll be working on this and willing to share my findings, but I'm unsure if my contribution will be useful because I have to achieve some results quickly and possibly I'll not stick to some IETF rules and standards!

For now I tried to use [signed_request](https://developers.facebook.com/docs/reference/login/signed-request) from Facebook. I had to add this in [XOAuthAuthenticationManager.getTokenFromCode()](https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/XOAuthAuthenticationManager.java#L388) (still doing some regression tests, but nothing seems broken for now):

```java
// Add missing client_id for Facebook request
body.add("client_id", config.getRelyingPartyId());
```

Of course, the Facebook non compliant details didn't stop there and I'll have a long road ahead!

When you're ready to share you branch, let me know. I'll be happy to give a try!

Thanks!

**Comment by bsekar** [comment]
This is now supported. Addressed by #630
